### ISSUE TITLE
Bug: crash on 404 when resource ID type does not match query string

### ISSUE THREAD

**Issue by bart-degreed** [json-api-dotnet/JsonApiDotNetCore#644]
## Description
See https://github.com/json-api-dotnet/JsonApiDotNetCore/issues/631#issuecomment-559562320

Looking at the changes, the trim/split/findindex url parsing feels a bit brittle to me. Wouldn't it be an option to use the route values, like it is done at:
https://github.com/json-api-dotnet/JsonApiDotNetCore/blob/d9dd76069a35bfbd8bb180ff83213014d2601b10/src/JsonApiDotNetCore/Middleware/CurrentRequestMiddleware.cs#L251?

For example, the debugger shows me for the request path:
```
api/v3/users/ABC/completions
```

these route values:
```
{
  "action": "GetRelationship"
  "controller": "Users"
  "id": "ABC"
  "relationshipName": "completions"
}
```

Hope this helps.

**Comment by wisepotato** [comment]
WOrking on this at the moment, found some weird things. I shouldn't code late at night or early in the morning. I added a relationshipId which should never be set. I'm refactoring the `CurrentRequestMiddleware` and the adjoining tests to use `_routeValues`.

Sorry for the inconvenience. Probably a fix incoming either today or monday.

**Comment by bart-degreed** [comment]
No worries, we all have days like that ;)

**Comment by bart-degreed** [comment]
@wisepotato I've been working from a detached head since this was introduced. Now that more of my colleagues are going to work on my prototype and I'm hoping to contribute PRs to this repo in the near future, this starts to get inconvenient. 

Can you maybe revert this commit until the proper solution is available?

**Comment by wisepotato** [comment]
I getcha, but will work on this tomorrow.

**Comment by maurei** [comment]
Got a few things to deliver before the end of the year, but I will prioritise this right after.

**Comment by wisepotato** [comment]
This is not yet closed.

**Comment by maurei** [comment]
Whoops, that was unintended

**Comment by bart-degreed** [comment]
Thanks, my issue is fixed by #651.

As a side effect, it also fixes a bug where the namespace part is missing when rendering URLs in links on getting a list of resources, which I had not noticed earlier.

Example: GET api/v3/articles/

Before:
```json
{
  "meta": {
    "total-records": 3
  },
  "links": {
    "self": "http://localhost/articles"
  },
  "data": [ ... ]
}
```

After
```json
{
  "meta": {
    "total-records": 3
  },
  "links": {
    "self": "http://localhost/api/v3/articles"
  },
  "data": [ ... ]
}
```
